metals/src/main/scala/scala/meta/internal/metals/CallHierarchyProvider.scala

```diff
@@ -0,0 +1,376 @@
+          new CallHierarchyItem(
+            info.displayName,
+            SymbolKind.Method,
+            source.toURI.toString,
+            range,
+            occurence.toLocation(source.toURI.toString).getRange
+          )
+        )
+      case SymbolInformation.Kind.CONSTRUCTOR | SymbolInformation.Kind.CLASS =>
+        Some(
+          new CallHierarchyItem(
+            info.displayName,
+            SymbolKind.Constructor,
+            source.toURI.toString,
+            range,
+            occurence.toLocation(source.toURI.toString).getRange
+          )
+        )
+      case _ =>
+        None
+    }
+    val symbol =
+      if (info.isClass)
+        info.signature
+          .asInstanceOf[s.ClassSignature]
+          .declarations
+          .flatMap(_.symlinks.find(_.endsWith("`<init>`().")))
+          .getOrElse(info.symbol)
+      else info.symbol
+
+    val displayName =
+      if (info.isConstructor && info.displayName == "<init>")
+        info.symbol.slice(
+          info.symbol.lastIndexOf("/") + 1,
+          info.symbol.lastIndexOf("#")
+        )
+      else info.displayName
+
+    chi.foreach(item => {
+      item.setData(CallHierarchyItemInfo(displayName, symbol))
+      item.setName(displayName)
+    })
+    chi
+  }
+
+  def prepare(params: CallHierarchyPrepareParams): List[CallHierarchyItem] = {
+    val source = params.getTextDocument.getUri.toAbsolutePath
+
+    semanticdbs.textDocument(source).documentIncludingStale match {
+      case Some(doc) =>
+        val results: List[ResolvedSymbolOccurrence] =
+          definition.positionOccurrences(source, params.getPosition, doc)
+        results.flatMap { result =>
+          for {
+            occurence <- result.occurrence
+            range <- occurence.range
+            tree <- trees.findLastEnclosingAt(source, range.toLSP.getStart)
+            definition <- findDefinition(tree)
+            chi <- symbolOccurenceToCallHierarchyItem(
+              source,
+              doc,
+              result.occurrence.get,
+              definition.pos.toLSP
+            )
+          } yield chi
+        }
+      case None =>
+        Nil
+    }
+  }
+
+  private def findIncomingCalls(
+      source: AbsolutePath,
+      doc: TextDocument,
+      root: Tree,
+      info: CallHierarchyItemInfo
+  ): List[(SymbolOccurrence, l.Range, List[l.Range])] = {
+    def search(
+        tree: Tree,
+        parent: Option[Name],
+        parentRange: Option[l.Range]
+    ): List[(SymbolOccurrence, l.Range, l.Range)] = tree match {
+      case name: Name
+          if (parent.isDefined && definition
+            .positionOccurrences(source, name.pos.toLSP.getEnd, doc)
+            .flatMap(_.occurrence)
+            .exists(occ =>
+              occ.symbol == info.symbol && occ.role == Role.REFERENCE
+            )) ||
+            (name.parent.map(_.is[Term.Apply]).getOrElse(false) && definition
+              .positionOccurrences(source, name.pos.toLSP.getEnd, doc)
```

Comment:
You could find all symbol references first and create a set out of it to jsut check here.